control_plane/attachment_service/src/persistence.rs

```diff
@@ -1,234 +1,334 @@
+//     }
+
+//     async fn load_or_new(path: &Utf8Path) -> Self {
+//         match Self::load(path).await {
+//             Ok(s) => {
+//                 tracing::info!("Loaded state file at {}", path);
+//                 s
+//             }
+//             Err(e)
+//                 if e.downcast_ref::<std::io::Error>()
+//                     .map(|e| e.kind() == std::io::ErrorKind::NotFound)
+//                     .unwrap_or(false) =>
+//             {
+//                 tracing::info!("Will create state file at {}", path);
+//                 Self {
+//                     tenants: HashMap::new(),
+//                     path: path.to_owned(),
+//                 }
+//             }
+//             Err(e) => {
+//                 panic!("Failed to load state from '{}': {e:#} (maybe your .neon/ dir was written by an older version?)", path)
+//             }
+//         }
+//     }
+// }
+
+use std::collections::{HashMap, HashSet};
+use std::env;
+use std::str::FromStr;
+
+use control_plane::attachment_service::{NodeAvailability, NodeSchedulingPolicy};
+use diesel::pg::PgConnection;
+use diesel::prelude::*;
+use diesel::Connection;
+use pageserver_api::shard::{ShardCount, ShardNumber, TenantShardId};
 use postgres_connection::parse_host_port;
 use serde::{Deserialize, Serialize};
-use utils::{
-    generation::Generation,
-    id::{NodeId, TenantId},
-};
-
-use crate::{node::Node, PlacementPolicy};
-
-/// Placeholder for storage.  This will be replaced with a database client.
+use utils::generation::Generation;
+use utils::id::{NodeId, TenantId};
+
+use crate::node::Node;
+
+/// ## What do we store?
+///
+/// The attachment service does not store most of its state durably.
+///
+/// The essential things to store durably are:
+/// - generation numbers, as these must always advance monotonically to ensure data safety.
+/// - Tenant's PlacementPolicy and TenantConfig, as the source of truth for these is something external.
+/// - Node's scheduling policies, as the source of truth for these is something external.
+///
+/// Other things we store durably as an implementation detail:
+/// - Node's host/port: this could be avoided it we made nodes emit a self-registering heartbeat,
+///   but it is operationally simpler to make this service the authority for which nodes
+///   it talks to.
+///
+/// ## Performance/efficiency
+///
+/// The attachment service does not go via the database for most things: there are
+/// a couple of places where we must, and where efficiency matters:
+/// - Incrementing generation numbers: the Reconciler has to wait for this to complete
+///   before it can attach a tenant, so this acts as a bound on how fast things like
+///   failover can happen.
+/// - Pageserver re-attach: we will increment many shards' generations when this happens,
+///   so it is important to avoid e.g. issuing O(N) queries.
+///
+/// Database calls relating to nodes have very low performance expectations, as the number
```

Comment:
What does "very low performance expectations" mean? "We don't need database calls relating to nodes to be fast"? Could it maybe be reformulated?

Reply:
I've clarified this comment a bit to explain that reads aren't important because we use in-memory state, whereas writes matter but have a low volume.